### ISSUE TITLE
Language generation problems

### ISSUE THREAD

**Issue by rusiaaman** [zihangdai/xlnet#94]
Thanks for making this opensource.

I am writing this specifically from the view point of text generation using XLNet. I have experimented with generating text, and the samples can be seen here https://github.com/rusiaaman/XLnet-gen/blob/master/samples/conditional_top-p-0.9.txt

Issues:
1. There has to be a long context for XLNet to generate meaningful samples. This is most probably because of the way XLNet was trained: 85 tokens out of 512 are predicted at a time. 

Fortunately, due to the presence of \<eod\> token, having a (grammatically correct) random paragraph followed by \<eod\> seems to give good results even for short prompts. But if we give no prompt, the random context influences the predictions a lot.

 2. Due to the way perm_mask was generated during pre-training, all context tokens can attend to each other. ~~During text generation, if we let the context tokens attend to only the past context tokens, the output is gibberish. On the other hand if let the context tokens attend to each other, the first sampled token will be of high quality.~~

~~Can the above two problems be reduced if in pretraining we have causal attention mask for context tokens too? Or a middle ground where a block of context tokens have bi-directional attention and the rest uni-directional?~~

3. ~~If we let all context tokens attend to each other, but not the sampled tokens, the output quickly degrades after 3-4 tokens. To rephrase, let there be 250 context tokens with bi-directional attention, followed by 3 sampled tokens with causal attention, the next sampled token will not be a good sample. But the same is not true if we sample *two* tokens, and in next iteration make them a part of the context with bi-directional attention, and then sample next two tokens, and so on. Then the samples are indistinguishable from the case where we make each new token a part of the context immediately (instead of two at a time).~~

~~This is probably because of the way n-grams are masked. On an average 2.2 consecutive tokens are masked (5 max n-grams, with inverse probability of generation) and are surrounded by 2.2*(alpha-1) = 11 context tokens. During text generation the scenario of more than 3-4 tokens being the target tokens (having causal attention masks) is not very friendly. And I suspect the quality degrades quickly as we sample more tokens due to a cascading effect.~~

~~Can this problem be solved if we have a initial block of context followed by continuous target-tokens?~~

---
I would like your thoughts on this, and if my understanding of the whole mechanism is correct.

**Comment by rusiaaman** [comment]
Closing this issue due to the discovery of a mistake in my code. Autogressively generating text leads to acceptable results now instead of gibberish.
